sql-plugin/src/main/scala/com/nvidia/spark/rapids/profiler.scala

```diff
@@ -0,0 +1,387 @@
+      enable()
+    }
+  }
+
+  private def startPollingDriver(): Unit = {
+    if (timerFuture.isEmpty) {
+      if (timer.isEmpty) {
+        timer = Some(TrampolineUtil.newDaemonSingleThreadScheduledExecutor("profiler timer"))
+      }
+      timerFuture = Some(timer.get.scheduleWithFixedDelay(() => try {
+        updateActiveFromDriver()
+      } catch {
+        case e: Exception =>
+          logError("Profiler timer task error: ", e)
+      }, driverPollMillis, driverPollMillis, TimeUnit.MILLISECONDS))
+    }
+  }
+
+  private def stopPollingDriver(): Unit = {
+    timerFuture.foreach(_.cancel(false))
+    timerFuture = None
+  }
+
+  private def updateActiveFromDriver(): Unit = {
+    writer.foreach { w =>
+      val (jobs, stages) = synchronized {
+        (activeJobs.toArray, activeStages.toArray)
+      }
+      val (completedJobs, completedStages, allDone) =
+        w.pluginCtx.ask(ProfileJobStageQueryMsg(jobs, stages))
+          .asInstanceOf[(Array[Int], Array[Int], Boolean)]
+      if (completedJobs.nonEmpty || completedStages.nonEmpty) {
+        synchronized {
+          completedJobs.foreach(activeJobs.remove)
+          completedStages.foreach(activeStages.remove)
+          if (activeJobs.isEmpty && activeStages.isEmpty) {
+            disable()
+            stopPollingDriver()
+          }
+        }
+      }
+      if (allDone) {
+        logWarning("No further jobs or stages to profile, shutting down")
+        shutdown()
+      }
+    }
+  }
+}
+
+class ProfileWriter(
+    val pluginCtx: PluginContext,
+    profilePathPrefix: String,
+    codec: Option[CompressionCodec]) extends Profiler.DataWriter {
+  val executorId: String = pluginCtx.executorID()
+  private val outPath = getOutputPath(profilePathPrefix, codec)
+  private val out = openOutput(codec)
+  private var isClosed = false
+
+  override def write(data: ByteBuffer): Unit = {
+    if (!isClosed) {
+      while (data.hasRemaining) {
+        out.write(data)
+      }
+    }
+  }
+
+  override def close(): Unit = {
+    if (!isClosed) {
+      isClosed = true
+      out.close()
+      pluginCtx.send(ProfileEndMsg(executorId, outPath.toString))
+    }
+  }
+
+  private def getAppId: String = {
+    val appId = pluginCtx.conf.get("spark.app.id", "")
+    if (appId.isEmpty) {
+      java.lang.management.ManagementFactory.getRuntimeMXBean.getName
+    } else {
+      appId
+    }
+  }
+
+  private def getOutputPath(prefix: String, codec: Option[CompressionCodec]): Path = {
+    val parentDir = new Path(prefix)
+    val suffix = codec.map(c => "." + TrampolineUtil.getCodecShortName(c.getClass.getName))
+      .getOrElse("")
+    new Path(parentDir, s"rapids-profile-$getAppId-$executorId.bin$suffix")
+  }
+
+  private def openOutput(codec: Option[CompressionCodec]): WritableByteChannel = {
+    val hadoopConf = pluginCtx.ask(ProfileInitMsg(executorId, outPath.toString))
+      .asInstanceOf[SerializableConfiguration].value
+    val fs = outPath.getFileSystem(hadoopConf)
```

Comment:
Can we have some kind of a follow on issue to make this a more generic service/API? We have a number of other debug tools that want to write data out. It would be nice if we were consistent in how we got the configs to do so, and it would be even nicer if we didn't have to worry about it for each debug operation being done.